cmd/gator/test/test.go

```diff
@@ -0,0 +1,246 @@
+/*
+Copyright © 2021 NAME HERE <EMAIL ADDRESS>
+
+*/
+package test
+
+import (
+	"bufio"
+	"encoding/json"
+	"fmt"
+	"os"
+	"path/filepath"
+
+	"github.com/open-policy-agent/frameworks/constraint/pkg/types"
+	"github.com/open-policy-agent/gatekeeper/pkg/gator"
+	"github.com/open-policy-agent/gatekeeper/pkg/gator/test"
+	"github.com/open-policy-agent/gatekeeper/pkg/util"
+	"github.com/spf13/cobra"
+	"gopkg.in/yaml.v3"
+	"k8s.io/apimachinery/pkg/apis/meta/v1/unstructured"
+)
+
+const (
+	examples = `  # test a manifest containing Kubernetes objects, Constraint Templates, and Constraints
+  gator test --filename="manifest.yaml"
+
+  # test a directory
+  gator test --filename="config-and-policies/"
+
+  # Use multiple inputs
+  gator test --filename="manifest.yaml" --filename="templates-and-constraints/"
+
+  # Receive input from stdin
+  cat manifest.yaml | gator test
+
+  # Output structured violations data
+  gator test --filename="manifest.yaml" --json`
+)
+
+var Cmd = &cobra.Command{
+	Use:     "test",
+	Short:   "test resources against templates and constraints",
+	Example: examples,
+	Run:     run,
+}
+
+var allowedExtensions = []string{".yaml", ".yml", ".json"}
+
+var (
+	flagFilenames []string
+	flagOutput    string
+)
+
+const (
+	flagNameFilename = "filename"
+	flagNameOutput   = "output"
+
+	stringJSON = "json"
+	stringYAML = "yaml"
+)
+
+func init() {
+	Cmd.Flags().StringArrayVarP(&flagFilenames, flagNameFilename, "f", []string{}, "a file or directory containing kubernetes resources.  Can be specified multiple times.  Cannot be used in tandem with stdin.")
+	Cmd.Flags().StringVarP(&flagOutput, flagNameOutput, "o", "", fmt.Sprintf("Output format.  One of: %s|%s.", stringJSON, stringYAML))
+}
+
+func run(cmd *cobra.Command, args []string) {
+	unstrucs, err := readSources(flagFilenames)
+	if err != nil {
+		errFatalf("reading: %v", err)
+	}
+	if len(unstrucs) == 0 {
+		errFatalf("no input data identified")
+	}
+
+	responses, err := test.Test(unstrucs)
+	if err != nil {
+		errFatalf("auditing objects: %v\n", err)
+	}
+	results := responses.Results()
+
+	// TODO (https://github.com/open-policy-agent/gatekeeper/issues/1787): Add
+	// `-ojson` and `-oyaml` flags
+	switch flagOutput {
+	case stringJSON:
+		b, err := json.MarshalIndent(results, "", "    ")
+		if err != nil {
+			errFatalf("marshaling validation json results: %s", err)
+		}
+		fmt.Print(string(b))
+	case stringYAML:
+		b, err := yaml.Marshal(results)
+		if err != nil {
+			errFatalf("marshaling validation yaml results: %s", err)
+		}
+		fmt.Print(string(b))
+	default:
+		if len(results) > 0 {
+			for _, result := range results {
+				fmt.Printf("Message: %q", result.Msg)
+			}
+		}
+	}
+
+	// Whether or not we return non-zero depends on whether we have a `deny`
+	// enforcementAction on one of the violated constraints
```

Comment:
we should def mention this in our docs

Reply:
I've added info to the docs for `gator test` and it includes this blurb.  PTAL!